### ISSUE TITLE
Exception: API Key and the authentication credential are from different projects

### ISSUE THREAD

**Issue by gguuss** [GoogleCloudPlatform/google-cloud-java#1405]
If you have multiple sets of valid credentials configured in your Java environment, the Cloud library will not resolve between the two.

## Repro steps

1. Configure client library for API key: `export GOOGLE_API_KEY=SOME_VALID_DEVELOPER_KEY`
2. Configure client library for valid credentials from separate project: `export GOOGLE_APPLICATION_CREDENTIALS=./creds.json`
3. Run java sample, e.g. from github.com/GoogleCloudPlatform/java-docs-samples/translate...

```
mvn clean compile assembly:single
JAR_FILE=target/translate-1.0-jar-with-dependencies.jar
INPUT="A quick brown fox jumped over a lazy dog."
SOURCE_LANG="en"
TARGET_LANG="fr"
java -jar $JAR_FILE translate "$INPUT" $SOURCE_LANG $TARGET_LANG
```

## Expected outcome

Client library warns developer but uses an access token from the service account.

## Actual outcome
Client library fails, throwing exception and crashing app.

    Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 400 Bad Request
    {
      "code" : 400,
      "errors" : [ {
        "domain" : "global",
        "message" : "The API Key and the authentication credential are from different projects.",
        "reason" : "badRequest"
      } ],
      "message" : "The API Key and the authentication credential are from different projects.",
      "status" : "INVALID_ARGUMENT"
    }

**Comment by garrettjonesgoogle** [comment]
We are seeing this in the AppVeyor integration test - looks like an important issue to address.

**Comment by lesv** [comment]
GOOGLE_APPLICATION_CREDENTIALS should override `gcloud auth application-default login`

**Comment by gguuss** [comment]
@lesv This is happening when you are using the GOOGLE_API_KEY environment variable alongside the GOOGLE_APPLICATION_CREDENTIALS variable. I am not sure whether the default application credentials are affecting the conflict.

**Comment by lesv** [comment]
It doesn't sound like this will block GA for Storage, Logging, PubSub, Datastore, or BigQuery

**Comment by lesv** [comment]
I think this should block GA for Translate.

**Comment by tcoffee-google** [comment]
@gguuss It shouldn't be hard to ignore GOOGLE_API_KEY when GOOGLE_APPLICATION_CREDENTIALS is specified, but I'm wondering about the case where an explicit API key is passed to a call --- in this case, would you expect the service account credentials to be ignored?

**Comment by tcoffee-google** [comment]
(This may also be relevant to proper resolution of #1594.)

**Comment by tcoffee-google** [comment]
Initial modification at https://github.com/GoogleCloudPlatform/google-cloud-java/pull/2151 is expected to resolve the specific error seen here, but I'd like to add some additional tests once we have consensus on proper behavior where an explicit API key is provided.

**Comment by garrettjonesgoogle** [comment]
Marking as release blocking until @tcoffee-google 's change #2151 is submitted.

**Comment by gguuss** [comment]
@tcoffee-google I'm in favor of ignoring API key when Google application credentials is specified and somewhere documenting precedence for credentials. Regarding API key, it would be helpful to have a scenario that shows what you mean when referring to explicitly including an API key with an API call.  Is there a way to call the client library and include the API key in the call?

**Comment by tcoffee-google** [comment]
@gguuss Sorry, I shouldn't have said "call" (this is done in the Apiary client library, which supports the API key query parameter with each request) --- I meant where the API key is explicitly specified in the service options, i.e., through [TranslateOptions.Builder.setApiKey](http://googlecloudplatform.github.io/google-cloud-java/0.20.0/apidocs/com/google/cloud/translate/TranslateOptions.Builder.html#setApiKey-java.lang.String-).

It seems to me a user would be very surprised if they explicitly set an API key in their service options but it was overridden by the `GOOGLE_APPLICATION_CREDENTIALS` environment variable. Thus I would say the order of precedence should be:

1. explicitly set service account credentials (through [ServiceOptions.Builder.setCredentials](http://googlecloudplatform.github.io/google-cloud-java/0.20.0/apidocs/com/google/cloud/ServiceOptions.Builder.html#setCredentials-com.google.auth.Credentials-)
2. explicitly set API key (through [TranslateOptions.Builder.setApiKey](http://googlecloudplatform.github.io/google-cloud-java/0.20.0/apidocs/com/google/cloud/translate/TranslateOptions.Builder.html#setApiKey-java.lang.String-))
3. default service account credentials (through `GOOGLE_APPLICATION_CREDENTIALS`)
4. default API key (through `GOOGLE_API_KEY`).

What do you think?

**Comment by gguuss** [comment]
> Thus I would say the order of precedence should be:
>
>1. explicitly set service account credentials (through ServiceOptions.Builder.setCredentials
>1. explicitly set API key (through TranslateOptions.Builder.setApiKey)
>1. default service account credentials (through GOOGLE_APPLICATION_CREDENTIALS)
>1. default API key (through GOOGLE_API_KEY).

@tcoffee-google I agree, environment configuration should not override explicit configuration (e.g. through a builder); your ordering makes sense to me.

**Comment by tcoffee-google** [comment]
Fixed by #2151.
